File: Python/OneLang/One/Transforms/UseDefaultCallArgsExplicitly.py
from onelang_core import *
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Types as types
import OneLang.One.AstTransformer as astTrans

class UseDefaultCallArgsExplicitly(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("UseDefaultCallArgsExplicitly")
    
    def get_new_args(self, args, method):
        if "UseDefaultCallArgsExplicitly" in method.attributes and method.attributes.get("UseDefaultCallArgsExplicitly") == "disable":
            return args
        if len(args) >= len(method.parameters):
            return args
        
        new_args = []
        i = 0
        
        while i < len(method.parameters):
            init = method.parameters[i].initializer
            if i >= len(args) and init == None:
                self.error_man.throw(f'''Missing default value for parameter #{i + 1}!''')
                break
            new_args.append(args[i] if i < len(args) else init)
            i = i + 1
        return new_args
    
    def visit_expression(self, expr):
        super().visit_expression(expr)
        if isinstance(expr, exprs.NewExpression) and expr.cls_.decl.constructor_ != None:
            expr.args = self.get_new_args(expr.args, expr.cls_.decl.constructor_)
        elif isinstance(expr, exprs.InstanceMethodCallExpression):
            expr.args = self.get_new_args(expr.args, expr.method)
        elif isinstance(expr, exprs.StaticMethodCallExpression):
            expr.args = self.get_new_args(expr.args, expr.method)
        return expr

File: PHP/src/OneLang/One/Transforms/UseDefaultCallArgsExplicitly.php
<?php

namespace OneLang\One\Transforms\UseDefaultCallArgsExplicitly;

use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\IMethodCallExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\Expressions\NewExpression;
use OneLang\One\Ast\Expressions\StaticMethodCallExpression;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\IMethodBaseWithTrivia;
use OneLang\One\Ast\Types\MethodParameter;
use OneLang\One\AstTransformer\AstTransformer;

class UseDefaultCallArgsExplicitly extends AstTransformer {
    function __construct() {
        parent::__construct("UseDefaultCallArgsExplicitly");
        
    }
    
    protected function getNewArgs($args, $method) {
        if (array_key_exists("UseDefaultCallArgsExplicitly", $method->attributes) && (@$method->attributes["UseDefaultCallArgsExplicitly"] ?? null) === "disable")
            return $args;
        if (count($args) >= count($method->parameters))
            return $args;
        
        $newArgs = array();
        for ($i = 0; $i < count($method->parameters); $i++) {
            $init = $method->parameters[$i]->initializer;
            if ($i >= count($args) && $init === null) {
                $this->errorMan->throw("Missing default value for parameter #" . ($i + 1) . "!");
                break;
            }
            $newArgs[] = $i < count($args) ? $args[$i] : $init;
        }
        return $newArgs;
    }
    
    protected function visitExpression($expr) {
        parent::visitExpression($expr);
        if ($expr instanceof NewExpression && $expr->cls->decl->constructor_ !== null)
            $expr->args = $this->getNewArgs($expr->args, $expr->cls->decl->constructor_);
        else if ($expr instanceof InstanceMethodCallExpression)
            $expr->args = $this->getNewArgs($expr->args, $expr->method);
        else if ($expr instanceof StaticMethodCallExpression)
            $expr->args = $this->getNewArgs($expr->args, $expr->method);
        return $expr;
    }
}
